firefox-ios/Client/Experiments/Messaging/GleanPlumbMessageManager.swift

```diff
@@ -188,38 +191,58 @@ class GleanPlumbMessageManager: GleanPlumbMessageManagerProtocol {
 
         guard let helper = createMessagingHelper.createNimbusMessagingHelper() else { return }
 
-        // Make substitutions where they're needed.
-        let template = message.action
-        let uuid = helper.getUuid(template: template)
-        let action = helper.stringFormat(template: template, uuid: uuid)
-
-        // Create the message action URL.
-        let urlString = action.hasPrefix("://") ? URL.mozInternalScheme + action : action
-        guard let url = URL(string: urlString, invalidCharacters: false) else {
+        guard let (uuid, urlToOpen) = try? self.generateUuidAndFormatAction(for: message, with: helper) else {
             self.onMalformedMessage(id: message.id, surface: message.surface)
             return
         }
 
-        var urlToOpen = url
-
-        // We open webpages using our internal URL scheme
-        if url.isWebPage(), var components = URLComponents(string: "\(URL.mozInternalScheme)://open-url") {
-            components.queryItems = [URLQueryItem(name: "url", value: url.absoluteString)]
-            urlToOpen = components.url ?? url
-        }
-
         // With our well-formed URL, we can handle the action here.
         applicationHelper.open(urlToOpen)
 
         var extras = baseTelemetryExtras(using: message)
-        extras[MessagingKey.actionUUID.rawValue] = uuid ?? "nil"
+        if let uuid = uuid {
+            extras[MessagingKey.actionUUID.rawValue] = uuid
+        }
         TelemetryWrapper.recordEvent(category: .action,
                                      method: .tap,
                                      object: .messaging,
                                      value: .messageInteracted,
                                      extras: extras)
     }
 
+    private func generateUuidAndFormatAction(for message: GleanPlumbMessage,
+                                             with helper: NimbusMessagingHelperProtocol) throws -> (String?, URL) {
+        // Make substitutions where they're needed.
+        let actionTemplate = message.action
+        var uuid = helper.getUuid(template: actionTemplate)
+        let action = helper.stringFormat(template: actionTemplate, uuid: uuid)
+        let actionParams = message.data.actionParams
+
+        let urlString: String = action.hasPrefix("://") ? deepLinkScheme + action : action
+        guard var components = URLComponents(string: urlString) else {
+            throw NimbusError.UrlParsingError(message: "\(urlString) is not a valid URL")
+        }
+        var queryItems = components.queryItems ?? []
+
+        for (k, v) in actionParams {
```

Comment:
```suggestion
        for (key, value) in actionParams {
```
We'll be adding a swiftlint rule to disallow variable names shorter than three characters, so I'm just avoiding adding new things to fix later on.